server/src/main/java/org/elasticsearch/action/search/CanMatchNodeRequest.java

```diff
@@ -146,6 +151,37 @@ public CanMatchNodeRequest(
         indices = shards.stream().map(Shard::getOriginalIndices).flatMap(Arrays::stream).distinct().toArray(String[]::new);
     }
 
+    private static void collectAggregationQueries(Collection<AggregationBuilder> aggregations, List<QueryBuilder> aggregationQueries) {
+        for (AggregationBuilder aggregation : aggregations) {
+            QueryBuilder aggregationQuery = aggregation.getQuery();
+            if (aggregationQuery != null) {
+                aggregationQueries.add(aggregationQuery);
+            }
+            collectAggregationQueries(aggregation.getSubAggregations(), aggregationQueries);
+        }
+    }
+
+    private SearchSourceBuilder getCanMatchSource(SearchRequest searchRequest) {
+        // Aggregations may use a different query than the top-level search query. An example is
+        // the significant terms aggregation, which also collects data over a background that
+        // typically much larger than the search query. To accommodate for this, we take the union
+        // of all queries to determine whether a request can match.
+        List<QueryBuilder> aggregationQueries = new ArrayList<>();
+        if (searchRequest.source() != null && searchRequest.source().aggregations() != null) {
+            collectAggregationQueries(searchRequest.source().aggregations().getAggregatorFactories(), aggregationQueries);
+        }
+        if (aggregationQueries.isEmpty()) {
+            return searchRequest.source();
+        } else {
+            BoolQueryBuilder query = QueryBuilders.boolQuery();
+            query.should(searchRequest.source().query());
+            for (QueryBuilder aggregationQuery : aggregationQueries) {
+                query.should(aggregationQuery);
+            }
+            return SearchSourceBuilder.searchSource().query(query);
```

Comment:
I have a bad feeling about replacing the source entirely. And I get anxious that there are no test failures in your PR :) 

If you look at the [`canRewriteToMatchNone`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/SearchService.java#L1755) , we use more than just the query from the source. 

I believe that if we had a test with the new functionality, in combination with aggs that must visit all docs, or a request that has a suggest section, we'd see failures. Perhaps the same with subsearches. It would be good to expand testing perhaps, and find a different way to adjust the query without rewriting the entire `SearchSourceBuilder`. A copy constructor would be handy perhaps. I am not sure that setting the query on the existing search source builder instance is a good idea, actually that will very likely break things.